File: LeetCode/Study Plans/I150/1D-DP/198.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  int rob(vector<int> &nums) {
    if (nums.size() == 1) {
      return nums[0];
    } else if (nums.size() == 2) {
      return max(nums[0], nums[1]);
    }

    vector<int> dp(nums.size() + 1, 0);
    dp[0] = 0;
    dp[1] = nums[0];

    auto max_value = [&]() -> int {
      for (int i = 1; i < nums.size(); i++) {
        dp[i + 1] = max(dp[i - 1] + nums[i], dp[i]);
      }
      return dp[nums.size()];
    };
    return max_value();
  }
};

int main() {
  Solution s;
  vector<int> nums1 = {1, 2, 3, 1};
  vector<int> nums2 = {2, 7, 9, 3, 1};
  vector<int> nums3 = {2, 1, 1, 2};

  cout << s.rob(nums3) << "\n";
}

File: Interview/Databricks/Leetcode/30d/198.py
class Solution:
    def rob(self, nums: List[int]) -> int:  
        if len(nums) == 1:
            return nums[0]     

        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]
        dp[1] = max(nums[1], nums[0])
        
        for i in range(2, n):
            dp[i] = max(nums[i] + dp[i-2], dp[i-1])
        
        print(dp)

        return max(dp)